File: config/src/artifact/vorpal_website.rs
use crate::artifact::SYSTEMS;
use anyhow::Result;
use indoc::formatdoc;
use vorpal_sdk::{
    artifact::{bun::Bun, get_env_key, step, Artifact, ArtifactSource},
    context::ConfigContext,
};

#[derive(Default)]
pub struct VorpalWebsite {}

impl VorpalWebsite {
    pub fn new() -> Self {
        Self::default()
    }

    pub async fn build(self, context: &mut ConfigContext) -> Result<String> {
        let bun = Bun::new().build(context).await?;
        let bun_bin = format!("{}/bin", get_env_key(&bun));

        let name = "vorpal-website";

        let source = ArtifactSource::new(name, ".")
            .with_includes(vec!["website".to_string()])
            .with_excludes(vec![
                "website/.astro".to_string(),
                "website/README.md".to_string(),
                "website/dist".to_string(),
                "website/node_modules".to_string(),
            ])
            .build();

        let step_script = formatdoc! {r#"
            pushd ./source/vorpal-website/website
            {bun_bin}/bun install
            {bun_bin}/bun run build
            cp -r dist/* $VORPAL_OUTPUT/
        "#};

        let steps = vec![
            step::shell(
                context,
                vec![bun.clone()],
                vec![
                    "ASTRO_TELEMETRY_DISABLED=1".to_string(),
                    format!("PATH={bun_bin}"),
                ],
                step_script,
                vec![],
            )
            .await?,
        ];

        Artifact::new("vorpal-website", steps, SYSTEMS.to_vec())
            .with_sources(vec![source])
            .build(context)
            .await
    }
}

File: sdk/go/cmd/vorpal/artifact/vorpal_website.go
package artifact

import (
	"fmt"

	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/artifact"
	api "github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/api/artifact"
	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/config"
)

func BuildVorpalWebsite(context *config.ConfigContext) (*string, error) {
	bun, err := artifact.Bun(context)
	if err != nil {
		return nil, fmt.Errorf("failed to build bun: %w", err)
	}

	bunBin := fmt.Sprintf("%s/bin", artifact.GetEnvKey(*bun))

	name := "vorpal-website"

	source := artifact.NewArtifactSource(name, ".").
		WithIncludes([]string{"website"}).
		WithExcludes([]string{
			"website/.astro",
			"website/README.md",
			"website/dist",
			"website/node_modules",
		}).Build()

	stepScript := fmt.Sprintf(`pushd ./source/vorpal-website/website
%s/bun install
%s/bun run build
cp -r dist/* $VORPAL_OUTPUT/
`, bunBin, bunBin)

	step, err := artifact.Shell(
		context,
		[]*string{bun},
		[]string{
			"ASTRO_TELEMETRY_DISABLED=1",
			fmt.Sprintf("PATH=%s", bunBin),
		},
		stepScript,
		nil,
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create shell step: %w", err)
	}

	return artifact.NewArtifact(name, []*api.ArtifactStep{step}, SYSTEMS).
		WithSources([]*api.ArtifactSource{&source}).
		Build(context)
}
